cou/apps/auxiliary.py

```diff
@@ -183,6 +183,87 @@ class RabbitMQServer(AuxiliaryApplication):
+        :type target: OpenStackRelease
+        :param units: Units to generate upgrade plan
+        :type units: Optional[list[Unit]]
+        :return: List of pre upgrade steps.
+        :rtype: list[PreUpgradeStep]
+        """
+        steps = super().pre_upgrade_steps(target, units)
+        if self.config["enable-auto-restarts"].get("value") is False:
+            action_name = "run-deferred-hooks"
+            units_to_run_action = self.units.values() if units is None else units
+            steps += [
+                PreUpgradeStep(
+                    description="Auto restarts is disabled, will"
+                    f" execute run-deferred-hooks for unit: '{unit.name}'",
+                    coro=self.model.run_action(unit.name, action_name, raise_on_failure=True),
+                )
+                for unit in units_to_run_action
+            ]
+            steps += [
+                PreUpgradeStep(
+                    description=(
+                        f"Wait for up to {self.wait_timeout}s for app '{self.name}'"
+                        " to reach the idle state"
+                    ),
+                    parallel=False,
+                    coro=self.model.wait_for_active_idle(self.wait_timeout, apps=[self.name]),
+                )
+            ]
+        return steps
+
+    def post_upgrade_steps(
+        self, target: OpenStackRelease, units: Optional[list[Unit]]
+    ) -> list[PostUpgradeStep]:
+        """Post Upgrade steps planning.
+
+        Wait until the application reaches the idle state and then check the target workload.
+
+        :param target: OpenStack release as target to upgrade.
+        :type target: OpenStackRelease
+        :param units: Units to generate post upgrade plan
+        :type units: Optional[list[Unit]]
+        :return: List of post upgrade steps.
+        :rtype: list[PostUpgradeStep]
+        """
+        steps = []
+        if self.config["enable-auto-restarts"].get("value") is False:
+            steps += [
+                PostUpgradeStep(
+                    description=(
+                        f"Wait for up to {self.wait_timeout}s for app '{self.name}'"
+                        " to reach the idle state"
+                    ),
+                    parallel=False,
+                    coro=self.model.wait_for_active_idle(self.wait_timeout, apps=[self.name]),
+                )
+            ]
+            action_name = "run-deferred-hooks"
+            units_to_run_action = self.units.values() if units is None else units
+            steps += [
+                PostUpgradeStep(
+                    description="Auto restarts is disabled, will"
+                    f" execute run-deferred-hooks for unit: '{unit.name}'",
+                    coro=self.model.run_action(unit.name, action_name, raise_on_failure=True),
+                )
+                for unit in units_to_run_action
+            ]
+        steps += super().post_upgrade_steps(target, units)
+        return steps
```

Comment:
I hope post upgrade make sense to you: since the auto restart is disabled, we need to run the deferred action and restart the service (originally done automatically by auto restart = True).

For pre upgrade: since auto restart is disabled, we don't know the if the service has pending events or not, so we want to run `run-deferred-hooks` action to clear the event before performing upgrade. If there are no pending events, this step should be a no-op, so it's safe to run anyway

Reply:
I think that it make sense to create a helper function which will take the type of step as argument, e.g. `PostUpgradeStep` or `PreUpgradeStep`

Reply:
that's does not help much though, they look similar but the order is different and important

Reply:
done

Reply:
You can use `[::-1]` to verse order, or create two helper function for each step. But I think it's fine now :+1: